### PR TITLE
`Development`: Provide standalone problem statement for external apps

### PR DESCRIPTION
**Pull request by sven0311**
FYI: This PR was already reviewed [here](https://github.com/ls1intum/Artemis/pull/6355) but reopened in this new PR since the CI does not work for forks.

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

The problem statement and the feedback view are currently to complex to move them to the native clients with a native implementation. Therefore, we want to embed them using web views.
### Description

We refactor the problem statement to be able to exist on its own. Furthermore, we add a new component that can show the feedback component based on query args.

### Steps for Testing
0. Deploy to TestServer2 (since the id's below are only valid for testServer 2, otherwise you can use another test server as well but research the exerciseID , participationID and resultId yourself)
1. Login with Artemis_admin account
2. Go to https://artemis-test2.artemis.in.tum.de/courses/567/exercises/1513/problem-statement/3031
3. Verify that you can see the problem statement end it looks fine (only works for light mode, dark mode is handled in native clients (background is transparent)
4. Go to https://artemis-test2.artemis.in.tum.de/courses/567/exercises/1513/participations/3031/results/3931/feedback
5. Verify that you can see the feedback to your result (only works for light mode, dark mode is handled in native clients (background is transparent)
6. Also verify that https://artemis-test2.artemis.in.tum.de/courses/567/exercises/1513/ still looks like it should

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| problem-statement.component.ts | 82% | 96% |
| standalone-feedback.component.ts | 77% | 98% |

### Screenshots
IMG_3584

IMG_3585

### CODE REVIEW

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts [hunk 1]
```diff
@@ -418,7 +418,7 @@ export class CourseExerciseDetailsComponent implements OnInit, OnDestroy {
         });
 
         if (this.exercise!.type === ExerciseType.MODELING || this.exercise!.type === ExerciseType.TEXT) {
-            return this.gradedStudentParticipation?.results?.find((result: Result) => !!result.completionDate) || undefined;
+            return;
```
Comment (Strohgelaender) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:421]: When you are already at it, you can specify the return type of this method in its signature. This way, errors like this get caught by the compiler.

File: src/main/webapp/app/overview/exercise-details/problem-statement/problem-statement.component.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    styleUrls: ['../../course-overview.scss'],
+})
+export class ProblemStatementComponent implements OnInit {
+    @Input()
+    public exercise?: Exercise;
+
+    @Input()
+    participation?: StudentParticipation;
+
+    constructor(public route: ActivatedRoute, private exerciseService: ExerciseService, private participationService: ParticipationService) {}
```
Comment (Strohgelaender) [inline src/main/webapp/app/overview/exercise-details/problem-statement/problem-statement.component.ts:21]: Is the route public on purpose?

File: src/main/webapp/app/overview/exercise-details/problem-statement/problem-statement.component.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    ngOnInit() {
+        this.route.params.subscribe((params) => {
+            const exerciseId = parseInt(params['exerciseId'], 10);
+            let participationId: number | undefined;
+            if (params['participationId']) {
+                participationId = parseInt(params['participationId'], 10);
+            } else {
+                participationId = undefined;
+            }
```
Comment (Strohgelaender) [inline src/main/webapp/app/overview/exercise-details/problem-statement/problem-statement.component.ts:31]: Not sure if  `let participationId?: number = undefined;` also works.
```suggestion
            let participationId: number | undefined = undefined;
            if (params['participationId']) {
                participationId = parseInt(params['participationId'], 10);
            }
```

File: src/main/webapp/app/shared/connection-warning/connection-warning.component.ts [hunk 4]
```diff
@@ -43,7 +43,7 @@ export class JhiConnectionWarningComponent implements OnInit, OnDestroy {
     }
 
     ngOnDestroy() {
-        this.websocketStatusSubscription.unsubscribe();
-        this.routerSubscription.unsubscribe();
+        this.websocketStatusSubscription?.unsubscribe();
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/connection-warning/connection-warning.component.ts:46]: Why is this needed? According to the type, this value should not be undefined. Please change the type if this is now the case.

File: src/main/webapp/app/overview/exercise-details/problem-statement/problem-statement.component.ts (hunk 3 above)
Comment (sven0311) [reply src/main/webapp/app/overview/exercise-details/problem-statement/problem-statement.component.ts]: No it doesn't, but I applied your other suggestion 👍

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS2, seems to be working as expected

Screenshot 2023-04-12 085236
Screenshot 2023-04-12 085257

**Review by simon-hng — APPROVED** [review]
Code

**Review by terlan98 — APPROVED** [review]
Reapproved. Code looks good

File: src/main/webapp/app/exercises/shared/feedback/standalone-feedback/standalone-feedback.component.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    styleUrls: ['./../feedback.scss'],
+})
+export class StandaloneFeedbackComponent implements OnInit {
+    exercise?: Exercise;
+    result?: Result;
+
+    isTemplateStatusMissing = false;
+    showMissingAutomaticFeedbackInformation = false;
+    messageKey?: string = undefined;
+    exerciseType?: ExerciseType = undefined;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/shared/feedback/standalone-feedback/standalone-feedback.component.ts:22]: Using `?:`initializes the attribute with undefined by default, this explicit initialization is not necessary. 
```suggestion
    messageKey?: string;
    exerciseType?: ExerciseType;
```

File: src/main/webapp/app/shared/notification/loading-notification/loading-notification.component.ts [hunk 6]
```diff
@@ -23,6 +23,6 @@ export class LoadingNotificationComponent implements OnInit, OnDestroy {
     }
 
     ngOnDestroy() {
-        this.loadingSubscription.unsubscribe();
+        this.loadingSubscription?.unsubscribe();
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/notification/loading-notification/loading-notification.component.ts:26]: similar change as the reverted other, why is this needed?

File: src/test/javascript/spec/component/shared/feedback/standalone-feedback.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                // mock exerciseCacheService
+                exerciseCacheService = fixture.debugElement.injector.get(ExerciseCacheService);
+                getLatestDueDateMock = jest.spyOn(exerciseCacheService, 'getLatestDueDate');
+                getLatestDueDateMock.mockReturnValue(of(latestDueDate));
+            });
+    });
+
+    it('should create', () => {
+        expect(component).toBeTruthy();
+    });
```
Comment (Strohgelaender) [inline src/test/javascript/spec/component/shared/feedback/standalone-feedback.component.spec.ts:60]: How much value is in a test like this? Can this even fail?

Comment (sven0311) [reply src/test/javascript/spec/component/shared/feedback/standalone-feedback.component.spec.ts]: I saw this in a tutorial and it should check if the component can be initialised at all. But since this is also covered by the other tests I removed these

**Review by terlan98 — DISMISSED** [review]
The changes look good. Reapproved

**Review by simon-hng — DISMISSED** [review]
Reapprove code

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested again on TS2, seems good again
Screenshot 2023-04-14 171700

**Review by Strohgelaender — DISMISSED** [review]
Thanks for adressing my code comments

File: src/main/webapp/app/shared/layouts/main/main.component.ts [hunk 8]
```diff
@@ -1,22 +1,28 @@
...
 import { SentryErrorHandler } from 'app/core/sentry/sentry.error-handler';
 import { ThemeService } from 'app/core/theme/theme.service';
+import { DOCUMENT } from '@angular/common';
 
 @Component({
     selector: 'jhi-main',
     templateUrl: './main.component.html',
 })
 export class JhiMainComponent implements OnInit {
+    public showSkeleton = true;
```
Comment (krusche) [inline src/main/webapp/app/shared/layouts/main/main.component.ts:15]: can we document the purpose of this attribute here in the code and mention standalone components shown in mobile devices? Please refer to the corresponding components

File: src/main/webapp/app/shared/layouts/main/main.component.ts [hunk 9]
```diff
@@ -49,4 +65,10 @@ export class JhiMainComponent implements OnInit {
 
         this.themeService.initialize();
     }
+
+    private shouldShowSkeleton(url: string): boolean {
```
Comment (krusche) [inline src/main/webapp/app/shared/layouts/main/main.component.ts:69]: I would also appreciate some code documentation here

File: src/main/webapp/app/shared/layouts/main/main.component.ts [hunk 10]
```diff
@@ -38,6 +44,16 @@ export class JhiMainComponent implements OnInit {
 
     ngOnInit() {
         this.router.events.subscribe((event) => {
+            if (event instanceof NavigationStart) {
+                const newShowSkeleton = this.shouldShowSkeleton(event.url);
+                if (!newShowSkeleton && this.showSkeleton) {
+                    this.renderer.addClass(this.document.body, 'transparent-background');
```
Comment (krusche) [inline src/main/webapp/app/shared/layouts/main/main.component.ts:50]: Is there no other way to implement this? This looks quite ugly to be honest

File: src/main/webapp/app/shared/layouts/main/main.component.ts [hunk 11]
```diff
@@ -38,6 +44,16 @@ export class JhiMainComponent implements OnInit {
 
     ngOnInit() {
         this.router.events.subscribe((event) => {
+            if (event instanceof NavigationStart) {
+                const newShowSkeleton = this.shouldShowSkeleton(event.url);
```
Comment (krusche) [inline src/main/webapp/app/shared/layouts/main/main.component.ts:48]: why do we need `newShowSkeleton` here? Please document the rationale in the code

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.module.ts [hunk 12]
```diff
@@ -40,7 +43,29 @@ const routes: Routes = [
...
+                path: 'problem-statement',
+                component: ProblemStatementComponent,
+            },
+            {
+                pathMatch: 'full',
+                path: 'problem-statement/:participationId',
+                component: ProblemStatementComponent,
+            },
+            {
+                path: 'feedback/:participationId/:resultId/:isTemplateStatusMissing',
```
Comment (krusche) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.module.ts:66]: I don't understand this URL, it does not follow best practices.
Instead, it should be something like `participations/:participationId/results/:resultId/feedback?isTemplateStatusMissing=:isTemplateStatusMissing`

I.e. `isTemplateStatusMissing` should rather be a query param (not sure if this is possible)

It's also unclear why `isTemplateStatusMissing` is needed and where it comes from.

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts [hunk 13]
```diff
@@ -401,7 +401,7 @@ export class CourseExerciseDetailsComponent implements OnInit, OnDestroy {
      * Returns the latest finished result for modeling and text exercises. It does not have to be rated.
      * For other exercise types it returns a rated result.
      */
-    getLatestRatedResult() {
+    getLatestRatedResult(): void {
```
Comment (krusche) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:404]: the methods name does not make sense. getSomething should not have a void return type.
Either return something or call it e.g. `loadLatestRatedResult`.

However it's also questionable that `this.complaintService.findBySubmissionId` is invoked here when "just the latest rated result should be retrieved. So in fact, this method rather loads the complaint based on the submission id for all exercises with an existing result and potentially updates the latest result for quiz or programming exercises.

When you touch the method, I think you should refactor it, because currently it is really bad code (not your fault) and the JSDoc is also horrible :-(

File: src/main/webapp/app/exercises/shared/feedback/standalone-feedback/standalone-feedback.component.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+@Component({
+    selector: 'jhi-standalone-feedback',
+    templateUrl: './standalone-feedback.component.html',
+    styleUrls: ['./../feedback.scss'],
+})
+export class StandaloneFeedbackComponent implements OnInit {
+    exercise?: Exercise;
+    result?: Result;
+
+    isTemplateStatusMissing = false;
```
Comment (krusche) [inline src/main/webapp/app/exercises/shared/feedback/standalone-feedback/standalone-feedback.component.ts:19]: documentation would be helpful here as well

**Review by krusche — CHANGES_REQUESTED** [review]
see my inline comments: code documentation is missing, client URLs do not follow best practices and the code quality of changed files should be improved.

File: src/main/webapp/app/shared/layouts/main/main.component.ts (hunk 8 above)
Comment (TimOrtel) [reply src/main/webapp/app/shared/layouts/main/main.component.ts:20]: I added clarifying comments to both the property and `shouldShowSkeleton`.

File: src/main/webapp/app/shared/layouts/main/main.component.ts (hunk 9 above)
Comment (TimOrtel) [reply src/main/webapp/app/shared/layouts/main/main.component.ts:89]: I added clarifying comments to both the property and `shouldShowSkeleton`.

File: src/main/webapp/app/shared/layouts/main/main.component.ts (hunk 10 above)
Comment (TimOrtel) [reply src/main/webapp/app/shared/layouts/main/main.component.ts:65]: I researched this and to my understanding there is no other way to overwrite the default background attibute of the body tag as it is outside of Angular's reach. We have to manually set a class on the background for this.

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.module.ts (hunk 12 above)
Comment (TimOrtel) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.module.ts]: Thanks for the suggestion. I have updated the route. Furthermore, I removed the `isTemplateStatusMissing` parameter as it is something that can be calculated by the web client itself.

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts (hunk 13 above)
Comment (TimOrtel) [reply src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts]: I renamed the function and the javadoc to better reflect the function's behavior.

File: src/main/webapp/app/exercises/shared/feedback/standalone-feedback/standalone-feedback.component.ts (hunk 14 above)
Comment (TimOrtel) [reply src/main/webapp/app/exercises/shared/feedback/standalone-feedback/standalone-feedback.component.ts]: The field has been removed.

**Review by terlan98 — COMMENTED** [review]
Tested on TS2. The URLs provided in `Steps for Testing` work fine. However, I noticed that now the problem statement doesn't render the markdown correctly when the user has not yet participated in the exercise.

image
URL: https://artemis-test2.artemis.in.tum.de/courses/8495/exercises/12508

**Review by terlan98 — APPROVED** [review]
Tested on TS2. Now works as expected. Thanks for implementing this 👍

**Review by FelixTJDietrich — DISMISSED** [review]
Tested in testing session on ts2. Works as intended.

**Review by pal03377 — APPROVED** [review]
Tested on TS2, works as expected 👍

**Review by lennart-keller — DISMISSED** [review]

**Review by pal03377 — DISMISSED** [review]
Still works fine after the merge conflict resolve, tested on TS2 in dark mode

**Review by terlan98 — DISMISSED** [review]
Retested on TS2 and reapproved 👍

**Review by FelixTJDietrich — DISMISSED** [review]
Tested again on ts2, still works as expected. Reapproved.

**Review by simon-hng — DISMISSED** [review]
Code looks good and understandable to me

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested again on TS2 all seems to be working as expected
